fs/wildcard.py

```diff
@@ -123,55 +135,92 @@ def get_matcher(patterns, case_sensitive):
     if not patterns:
         return lambda name: True
     if case_sensitive:
-        return partial(match_any, patterns)
+        return lambda name: match_any(patterns, name, accept_prefix)
     else:
-        return partial(imatch_any, patterns)
+        return lambda name: imatch_any(patterns, name, accept_prefix)
+
+
+class _PatternMatcher:
+    def __init__(self, pattern, case_sensitive, accept_prefixes):
+        self.case_sensitive = case_sensitive
+        self.accept_prefixes = accept_prefixes
+
+        self.tokens = []
+        i = 0
+        while i < len(pattern):
+            if pattern[i] == "[":
+                start = i
+                i = pattern.find("]", i) + 1
+                self.tokens.append(_PatternMatcherToken(pattern[start:i]))
+            elif pattern[i] == "]":
+                raise ValueError(
+                    pattern + " is not a valid wildcard pattern. (unmatched ])"
+                )
+            elif pattern[i] == "?":
+                self.tokens.append(_PatternMatcherToken("?"))
+                i += 1
+            elif pattern[i] == "*":
+                asterik_len = 1
+                while (asterik_len + i) < len(pattern) and pattern[
+                    asterik_len + i
+                ] == "*":
+                    asterik_len += 1
+                if asterik_len > 2:
+                    raise ValueError(
+                        pattern
+                        + " is not a valid wildcard pattern. "
+                        + "(sequence of three or more *)"
+                    )
+                self.tokens.append(_PatternMatcherToken("*" * asterik_len))
+                i += asterik_len
+            else:
+                self.tokens.append(_PatternMatcherToken(pattern[i]))
+                i += 1
+
+    def __call__(self, text):
+        return self._match(self.tokens, text)
+
+    def _match(self, tokens, text):
+        if not tokens:
+            return text == ""
+        if not text:
+            return self.accept_prefixes
+
+        match_count = tokens[0].match(text[0], self.case_sensitive)
+        if match_count == 0:
+            return False
+        elif match_count == 1:
+            return self._match(tokens[1:], text[1:])
+        else:
+            return self._match(tokens[1:], text[1:]) or self._match(tokens, text[1:])
 
 
-def _translate(pattern, case_sensitive=True):
-    # type: (Text, bool) -> Text
-    """Translate a wildcard pattern to a regular expression.
+class _PatternMatcherToken:
+    def __init__(self, token_str):
+        self.token_str = token_str
 
-    There is no way to quote meta-characters.
+    def __repr__(self):
+        return "_PatternMatcherToken (" + self.token_str + ")"
```

Comment:
Use a repr string here:
```python
return "_PatternMatcherToken({!r})".format(self.token_str)
```
this will make `repr` roundtrip with `eval`, which is good practice for simple objects.

Reply:
Thanks. TIL